NOISSUE - Make MQTT Broker Configurable

rodneyosodo opened the pull request:
Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

### What does this do?
This PR makes the MQTT broker configurable either as VerneMQ or NATS. NATS is a lightweight and highly performant message broker that offers MQTT support, making it a suitable replacement for VerneMQ in this context.

### Which issue(s) does this PR fix/relate to?
No issue

### List any changes that modify/break current functionality
1. Add docker profiles to switch between different modes. The available profiles are:
	- `vernemq_nats`: Uses `vernemq` as MQTT_BROKER and `nats` as MESSAGE_BROKER.
	- `vernemq_rabbitmq`: Uses `vernemq` as MQTT_BROKER and `rabbitmq` as MESSAGE_BROKER.
	- `nats_nats`: Uses `nats` as both MQTT_BROKER and MESSAGE_BROKER.
	- `nats_rabbitmq`: Uses `nats` as MQTT_BROKER and `rabbitmq` as MESSAGE_BROKER.

3. Integration of NATS as the MQTT broker.

### Have you included tests for your changes?
Tested manually

### Did you document any new/modified functionality?
https://github.com/mainflux/docs/pull/159

### Notes
To be merged after https://github.com/mainflux/mainflux/pull/1903 and https://github.com/mainflux/mainflux/pull/1902

- NATS follows MQTT v3.1.1 specification
- Limitations regarding Nats MQTT support can be found [here](https://docs.nats.io/running-a-nats-service/configuration/mqtt#limitations) 

#### How to test

1. NATS as MQTT Broker and NATS as Message Broker

	```bash
	MF_MESSAGE_BROKER_TYPE=nats MF_MQTT_BROKER_TYPE=nats make run up args="-d"
	MF_MESSAGE_BROKER_TYPE=nats MF_MQTT_BROKER_TYPE=nats make run_addons up args="-d"
	```

2. VerneMQ as MQTT Broker and NATS as Message Broker

	```bash
	MF_MESSAGE_BROKER_TYPE=nats MF_MQTT_BROKER_TYPE=vernemq make run up args="-d"
	MF_MESSAGE_BROKER_TYPE=nats MF_MQTT_BROKER_TYPE=vernemq make run_addons up args="-d"
	```

3. NATS as MQTT Broker and RabbitMQ as Message Broker

	```bash
	MF_MESSAGE_BROKER_TYPE=rabbitmq MF_MQTT_BROKER_TYPE=nats make run up args="-d"
	MF_MESSAGE_BROKER_TYPE=rabbitmq MF_MQTT_BROKER_TYPE=nats make run_addons up args="-d"
	```

4. VerneMQ as MQTT Broker and RabbitMQ as Message Broker

	```bash
	MF_MESSAGE_BROKER_TYPE=rabbitmq MF_MQTT_BROKER_TYPE=vernemq make run up args="-d"
	MF_MESSAGE_BROKER_TYPE=rabbitmq MF_MQTT_BROKER_TYPE=vernemq make run_addons up args="-d"
	```

rodneyosodo committed:
Minor changes on mqtt publisher using nats

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Remove vernemq dependencies

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Restore VerneMQ config files

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Fix Makefile to support custom Docker profiles

The Makefile has been updated to support custom Docker profiles.
Previously, the Makefile only supported the default profiles for the
message broker and MQTT broker. Now, the Makefile allows for custom
profiles to be specified using environment variables. If the
MF_BROKER_TYPE or MF_MQTT_BROKER_TYPE variables are not set, the
default values "nats" and "nats" will be used, respectively. This
change enables more flexibility in configuring the Docker environment
for the project.

The `run` target has also been modified to use the correct broker
configuration file based on the MF_BROKER_TYPE variable. The sed
command in the `run` target now replaces the placeholder in the
docker/docker-compose.yml file with the appropriate broker
configuration file.

This commit improves the Makefile to support custom Docker profiles
and ensures the correct broker configuration file is used when
running the project.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Fix queue binding issue in RabbitMQ pubsub

The commit fixes an issue in the RabbitMQ pubsub implementation where the queue binding was not correctly set up. Instead of using the topic as the queue name, the commit now uses a unique client ID generated by combining the topic and subscriber ID. This ensures that each subscriber has its own dedicated queue. The commit also updates the queue binding to use the correct queue name.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Refactor Docker config editing in Makefile

The commit refactors the `edit_docker_config` function in the Makefile to improve readability and maintainability. The changes include:

- Removing unnecessary conditionals related to the `rabbitmq` broker

These changes ensure that the Docker configuration is correctly updated based on the specified MQTT broker type.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Fix failing tests on RabbitMQ

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Refactor MQTT_BROKER comment in docker-compose.yml

The MQTT_BROKER comment in the docker-compose.yml file has been updated to provide a more accurate description of its functionality. The comment now states that the MQTT_BROKER handles MQTT communication between MQTT adapters and the message broker, instead of Mainflux services. This change improves clarity and aligns with the actual purpose of the MQTT_BROKER.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Rename `MF_BROKER` to `MF_MESSAGE_BROKER`

The Makefile and Semaphore configuration files have been refactored to update the variable names related to the message broker type.

These changes ensure consistency and clarity in the codebase by using more descriptive variable names related to the message broker type.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Fix Docker profile configuration for nats_rabbitmq

Update the Docker profile configuration for nats_rabbitmq by replacing the NATS URL in the .env file with the correct value.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Rename MF_BROKER_URL to MF_MESSAGE_BROKER_URL

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Fix MQTT QoS level in pubsub.go

The MQTT QoS level in the pubsub.go file was set to 1, which is the
default level. However, since NATS supports up to QoS 1, I updated the
QoS level comment to reflect this.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

SammyOina commented on pkg/messaging/rabbitmq/pubsub_test.go:
```diff
@@ -419,8 +419,6 @@ func TestPubSub(t *testing.T) {
 			assert.Equal(t, expectedMsg.Channel, receivedMsg.Channel, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, &expectedMsg, receivedMsg))
 			assert.Equal(t, expectedMsg.Created, receivedMsg.Created, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, &expectedMsg, receivedMsg))
 			assert.Equal(t, expectedMsg.Protocol, receivedMsg.Protocol, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, &expectedMsg, receivedMsg))
-			assert.Equal(t, expectedMsg.Publisher, receivedMsg.Publisher, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, &expectedMsg, receivedMsg))
```
why are these removed

rodneyosodo replied:
The published message only contains `channel` and `payload`

SammyOina commented on mqtt/handler.go:
```diff
@@ -55,19 +55,19 @@ var channelRegExp = regexp.MustCompile(`^\/?channels\/([\w\-]+)\/messages(\/[^?]
 
 // Event implements events.Event interface.
 type handler struct {
-	publishers []messaging.Publisher
```
why remove support for mutliple publishers

rodneyosodo replied:
It was not being utilized. We only used passed one publisher.

SammyOina commented on pkg/messaging/mqtt/pubsub.go:
```diff
@@ -18,7 +18,7 @@ import (
 const (
 	username = "mainflux-mqtt"
-	qos      = 2
+	qos      = 1 // At least once - Since NATS supports up to QoS 1.
```
should be configurable with each profile or via env or some other way to pass as an argument

rodneyosodo committed:
Refactor NewPublisher to accept QoS parameter

The NewPublisher function in the pkg/messaging/mqtt/publisher.go file has been refactored to accept a new parameter, qos, which represents the Quality of Service level for MQTT message publishing. This change allows for more flexibility in configuring the MQTT publisher.

The NewPublisher function now has the following signature:

```go
func NewPublisher(address string, qos uint8, timeout time.Duration) (messaging.Publisher, error)
```

This change ensures that the MQTT publisher can be created with the desired QoS level, enhancing the reliability and delivery guarantees of the published messages.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Fix test assertions in pubsub_test.go

The test assertions in the pubsub_test.go file were incorrect. This commit fixes the assertions to properly compare the expected and received message values.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>

rodneyosodo committed:
Test configurable MQTT broker

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

drasko reviewed (approved):
LGTM
